### PR TITLE
utils::{TopicPartitionOffset, PartitionOffset} have Result fields

### LINKED ISSUE #128: utils::{TopicPartitionOffset, PartitionOffset} have Result fields
Currently, the `utils::{TopicPartitionOffset, PartitionOffset}` structs' offset fields are `Result`s. In my opinion, it's kind of strange to have a struct with a `Result` in the middle of it for anything but maybe custom `Error` types. It can make error handling on the user-side really inconvenient.

Additionally, it's not entirely clear to me why `TopicPartitionOffset` is needed at all—it seems that it exists only to flatten responses, which leads to [lots of inefficient cloning](https://github.com/spicavigo/kafka-rust/blob/master/src/client/mod.rs#L1340) of data that it already had. I think in an instance like the linked one, it would be more appropriate to return a map which maps topics to `PartitionOffset`s; that way, only a move of one string is necessary, rather than a clone for every single offset response.

What do you think? I tried to start by changing the `Result`s to just their underlying type, but I can see it's a non-trivial refactor, so I thought I'd get your input first.

### PR DESCRIPTION
**Pull request by dead10ck**
Resolves #128

- [x] Make `KafkaClient.fetch_group_(topic_)?offsets` return a `Result<HashMap<String, Vec<PartitionOffset>>>`
- [x] Make `PartitionOffset.offset` an `i64` rather than `Result<i64>`
- [x] Change `KafkaClient::produce_messages` to return a new struct, `ProduceConfirm`, rather than `TopicPartitionOffset`, and remove `TopicPartitionOffset`

### CODE REVIEW

**Comment by dead10ck** [comment]
I didn't see any unit tests associated with these methods, so I didn't add any either. Is this just stuff that doesn't lend itself well to unit testing? Or do you have some kind of facility to mock a Kafka broker?

**Comment by xitep** [comment]
No, I don't think there are any tests for these methods. You can capture the datastream and dump it to a file for example (the code in `client::__get_response` might be helpful.) Exposing a (public) method to dump the raw binary data would be very helpful, though, I have no particular idea how to go about it at the moment.

File: src/client/mod.rs [hunk 1]
```diff
@@ -1366,7 +1367,10 @@ fn __fetch_group_offsets(req: protocol::OffsetFetchRequest,
                     }
                 }
             }
+
+            topic_map.insert(tp.topic, partition_offsets);
```
Comment (xitep) [inline src/client/mod.rs]: this - getting rid of the original clone/allocation - is really cool! looking forward to reviewing the whole change. :+1:

**Comment by dead10ck** [comment]
@xitep so I'm looking over the `KafkaClient::produce_messages` and related code for getting rid of the `TopicPartitionOffset` object. However, I'm seeing we already have an object which represents a `ProduceResponse`; is there any reason we can't just return that?

Additionally, as for changing the offset field to `Result<i64, KafkaCode>`, I'm actually not sure it makes sense to do that. In a produce response, the offset has meaning—specifically, ["The offset assigned to the first message in the message set appended to this partition."](https://cwiki.apache.org/confluence/display/KAFKA/A+Guide+To+The+Kafka+Protocol#AGuideToTheKafkaProtocol-ProduceResponse) So, even if there is a non-zero error code, this offset is still valid data that the user should be able to access.

Tangentially, while looking over the code, I notice that [when sending messages via `KafkaClient::produce_messages`](https://github.com/spicavigo/kafka-rust/blob/master/src/client/mod.rs#L1186-L1200), the code currently creates one `ProduceRequest` for every message; however, the Kafka API appears to [allow sending multiple messages inside a single `ProduceRequest`](https://cwiki.apache.org/confluence/display/KAFKA/A+Guide+To+The+Kafka+Protocol#AGuideToTheKafkaProtocol-ProduceRequest); I'm not sure if there is a limit to how big a request can be, but it seems wasteful to make a network request for each message if the API allows sending bulk requests.

**Comment by xitep** [comment]
Allow me to start in reverse order regarding your points:

* Regarding [sending multiple messages via KafkaClient::produce_messages](https://github.com/spicavigo/kafka-rust/blob/master/src/client/mod.rs#L1186-L1200) ... I think it's doing the right thing. The code you've highlighted is taking all of the messages to be send and merely "builds up" a _single_ `protocol::ProduceRequest` for each broker currently hosting the corresponding partition (note the creation of the `protocol::ProduceRequest` is in a closure to [HashMap::or_insert_with](https://doc.rust-lang.org/collections/btree_map/enum.Entry.html#method.or_insert_with) which gets called only once per host, the HashMap key). Only after this `HashMap<"host", ProduceRequest>` is fully build up, each of the requests is send to the corresponding host.

* Kafka guarantees that writing a "message set" to a kafka partition either succeeds as a whole or it doesn't (and that the messages of this message set will have subsequent offsets). That's why it makes sense for kafka to deliver only the offset of the first message of the message set sent (since clients know how many messages they sent as part of the message set they can calculate the offsets of the remaining messages in the message set and there is no need for kafka to return a more complicated response structure than it does now.) Given this and the reasoning in my previous point that `kafka-rust` indeed does send multiple messages in one go to the broker, the response (for a single partition) we can await here is either a valid offset indicating that all the data sent has been successfully acknowledged by the brokers or that it failed. Hence, in the error case for a particular partition, there is no offset to be read. That's what justifies the `Result<i64, error>` for me.

* The reason `KafkaClient::produce_messages` doesn't expose `protocol::ProduceResponse` directly is due to the `ProduceResponse` being a crate-private type within kafka-rust; it's nowhere in the public API of `kafka-rust`. In my opinion this is good since it gives us the possibility to change the produce-response type without breaking client code. As I already mentioned earlier, the result type of `produce_messages` will need to be changed in future, such that messages which didn't make it to kafka due to some error on a partition will be passed back to the client, who then can decide what to do about the situation (e.g. re-send the data.) Such changes are not related at all to protocol parsing which is the main responsibility of `ProduceReponse`. On the other hand we just could expose it. I did so for the types of the "fetch messages" method. However, it was substantial effort to keep implementation details private such that future code changes here wouldn't break clients needlessly. I think this effort is not justified for the "produce messages" response type.

**Comment by dead10ck** [comment]
- Oohhh, ok, my apologies, I misunderstood the code; I see now it only produces one message for each host.
- Hmm, so if there is an error, does the broker not send an offset at all? Perhaps I misunderstood the documentation as well. I took it to mean that even in the case of an error, it would send the offset that it *would have* assigned. If it doesn't send it at all, then yes, a `Result` does make sense.
- Oh, I see, that makes sense. Ok, I'll proceed with the `ProduceConfirm` object, then.

My apologies for all the questions and misunderstandings. I'm still new to Rust (in the practical sense), and the Kafka protocol.

**Comment by xitep** [comment]
* In the error case, kafka has to send some offset (since it has to fill the 8 bytes in the response structure). However, I'm not sure what value it actually sends. But I doubt it can be useful to us in this situation.
* No need to apologize for anything. I'm glad I can be helpful. (btw: I also feel new to Rust, since i'm practicing it merely in my free time of which there is not much.)

**Comment by dead10ck** [comment]
@xitep ok, I think this is ready for review

**Comment by xitep** [comment]
it seems i'll be able to review it over this weekend. thanks so far!

File: src/utils.rs [hunk 2]
```diff
@@ -2,20 +2,10 @@
...
 
-use error::Result;
-
 /// A retrieved offset for a particular partition in the context of an
 /// already known topic.
 #[derive(Debug)]
 pub struct PartitionOffset {
-    pub offset: Result<i64>,
-    pub partition: i32,
-}
```
Comment (xitep) [inline src/utils.rs]: this used to be what now is replaced by `ProduceTopicPartitionConfirm`: as suspected the size of the struct got smaller!

```rust

#[test]
fn test_size() {
    use std::mem;
    println!("new version: {}", mem::size_of::<ProduceTopicPartitionConfirm>());
    println!("old version: {}", mem::size_of::<PartitionOffset>());
}
```

```
new version: 24
old version: 56
```

:cool:

File: src/producer.rs [hunk 3]
```diff
@@ -254,20 +257,19 @@ impl<P: Partitioner> Producer<P> {
...
-        } else {
             assert_eq!(1, rs.len());
-            rs.pop().unwrap().offset.map(|_| ())
+            rs.pop().unwrap();
         }
+
+        Ok(())
     }
 
     /// Synchronously send all of the specified messages to Kafka.
```
Comment (xitep) [inline src/producer.rs]: Can you please add a _"To validate that all of the specified records have been successfully delivered, inspection of the offsets on the returned confirms is necessary."_ here?

File: src/client/mod.rs [hunk 4]
```diff
@@ -14,14 +14,14 @@ use std::time::{Duration, Instant};
...
 pub use utils::PartitionOffset;
-pub use utils::TopicPartitionOffset;
 
 #[cfg(feature = "security")]
 pub use self::network::SecurityConfig;
 
 use codecs::{ToByte, FromByte};
 use error::{Result, Error, KafkaCode};
 use protocol::{self, ResponseParser};
+use producer::ProduceConfirm;
```
Comment (dead10ck) [inline src/client/mod.rs]: Sure, will do.

File: src/client/mod.rs [hunk 5]
```diff
@@ -827,7 +827,7 @@ impl KafkaClient {
             for tp in resp.topic_partitions {
                 let e = res.entry(tp.topic).or_insert(vec![]);
                 for p in tp.partitions {
-                    e.push(p.into_offset());
+                    e.push(try!(p.into_offset()));
```
Comment (dead10ck) [inline src/client/mod.rs]: Yeah, this was part of our earlier discussion. I could certainly add another error type, provided you don't mind the extra time it will take to get this PR merged. I personally lean toward just doing it right the first time, but I'm also ok with just adding the debug line and adding the error type in a future PR. I leave it up to you :)

File: src/client/mod.rs [hunk 6]
```diff
@@ -1364,9 +1356,13 @@ fn __fetch_group_offsets(req: protocol::OffsetFetchRequest,
                         // ~ immeditaly abort with the error
                         return Err(e);
                     }
+
```
Comment (dead10ck) [inline src/client/mod.rs]: Whoops, yep, will do.

File: src/lib.rs [hunk 7]
```diff
@@ -41,13 +41,4 @@ mod protocol;
...
-trait KafkaClientInternals {
-    fn internal_produce_messages<'a, 'b, I, J>(&mut self,
-                                               ack: client::RequiredAcks,
-                                               ack_timeout: i32,
-                                               messages: I)
-                                               -> Result<Vec<client::TopicPartitionOffset>>
-        where J: AsRef<client::ProduceMessage<'a, 'b>>,
-              I: IntoIterator<Item = J>;
-}
+pub use client_internals::KafkaClientInternals;
```
Comment (dead10ck) [inline src/lib.rs]: Oh, I see; yeah, I just saw that this was a duplicate, but I failed to see it wasn't public. I'll just remove this line.

File: src/client/mod.rs (hunk 5 above)
Comment (xitep) [inline src/client/mod.rs]: feel free to get it right. and take the time it requires. there's certainly no pressure to get this out as soon as possible. fortunately, this is just for fun :)

File: src/producer.rs [hunk 8]
```diff
@@ -254,20 +257,19 @@ impl<P: Partitioner> Producer<P> {
...
             // ~ with no required_acks we get no response and
             // consider the send-data request blindly as successful
-            Ok(())
-        } else {
             assert_eq!(1, rs.len());
-            rs.pop().unwrap().offset.map(|_| ())
+            rs.pop().unwrap();
         }
+
+        Ok(())
```
Comment (dead10ck) [inline src/producer.rs]: Ohh, I see now. I'm sorry, I think I just had a brain fart when I read this function and didn't see what the `map(|_| ())` was doing. I'll fix this.

File: src/producer.rs [hunk 9]
```diff
@@ -631,6 +633,29 @@ impl<H: BuildHasher> Partitioner for DefaultPartitioner<H> {
...
+#[derive(Debug)]
+pub struct ProduceTopicPartitionConfirm {
+    /// The offset assigned to the first message in the message set appended
+    /// to this partition, or an error if one occurred.
+    pub offset: std::result::Result<i64, error::KafkaCode>,
+
+    /// The partition to which the message(s) were appended.
+    pub partition: i32,
+}
+
```
Comment (dead10ck) [inline src/producer.rs]: Yeah, I think either would work. `ProduceConfirm` and `ProducePartitionConfirm` seems fine.

File: src/protocol/produce.rs [hunk 10]
```diff
@@ -257,31 +257,35 @@ pub struct PartitionProduceResponse {
...
     }
 }
 
 impl TopicPartitionProduceResponse {
-    pub fn get_response(&self, topic: String) -> Vec<TopicPartitionOffset> {
-        self.partitions
+    pub fn get_response(self) -> ProduceConfirm {
+        let confirms = self.partitions
             .iter()
-            .map(|ref p| p.get_response(topic.clone()))
```
Comment (dead10ck) [inline src/protocol/produce.rs]: You think it's ok to make this method consuming? I did it to get rid of the need to clone, but I wasn't sure what you'd think.

File: src/producer.rs (hunk 8 above)
Comment (dead10ck) [inline src/producer.rs]: Hmm, I did just have a thought, though. It seems `Producer::send_all` returns the response from the broker, while `Producer::send` just returns `()`. What if the user would want to access the information in the returned `ProduceConfirm`, such as to know the offset of the new message? Do you think it would make sense to return a `Result<ProduceConfirm>` instead?

Comment (xitep) [inline src/producer.rs]: Well, yes, it would make sense. However, I consider `Producer` to be a higher-level API exposing less of the underlying details, therefore my first reaction would rather be to make `Producer::send_all` return a `Result<()>` instead (however, this would first require proper error handling within the `Producer` implementation.)

For some time, I'm playing with the idea to introduce some (user configurable) buffering into `Producer` such that `Producer::send/send_all` would first build up batches of message sets and send the records eventually only when some condition is met (e.g. the buffer gets full and/or some timeout is hit.) The idea is to give clients an easy API to work with, while allowing to send messages in batches - if they arrive quickly enough - which of course is much more efficient than sending them one by one. Such a feature would require that the `send/send_all` methods more or less do not return any of the "partition or offset information" returned by the broker.

At this point in time I have no preference and I'm fine with `Producer::send` returning `Result<ProduceConfirm>` now. I think it will be consistent with the `send_all` method.

Comment (dead10ck) [inline src/producer.rs]: Actually, I hadn't considered the case where acknowledgement isn't required. I'll just leave the return type as is for now and fix it as you suggested.

**Comment by dead10ck** [comment]
Ok, I think that takes care of all the items in your review--please let me know if I missed anything. I am leaving individual commits as is for now to make your review easier, but I will rebase before this gets merged.

I would like your opinion on the new error type and how I handled using it in `KafkaClient::fetch_offsets`, though. I figured the error type needs to have the topic string, but I wanted to avoid unneccessary cloning in the case where the response contained an error (see `PartitionOffsetResponse::into_offset`); however, the map `entry` API takes ownership of the key you pass it, so the only way I could think of to avoid cloning was to remove the topic's key and reinsert at the end every iteration; I figured this is still preferrable to cloning the topic `String` on every iteration. I know this is somewhat clunky though; do you see a better way to do it, or a change in the new error type that would make this easier?

**Comment by xitep** [comment]
thanks for the update! this looks all good except the unfortunate `hashmap#remove/insert` code in the `KafkaClient::fetch_offsets` method (in the error case I still see the topic being cloned further down the call chain in `PartitionOffsetResponse::into_offset`.)

i think the troubles are merely introduced by having the code being split between the `fetch_offsets` and `into_offset` methods. i think we can combine these two methods if it makes life easier for us (the only reason for the current split i see so far is readability, but it should not come at a cost of runtime overhead that we're trying to resolve.)

given the "old" `into_offset` implementation, the following "hacky" code in `fetch_offsets` should be doing what we're after (haven't tested, but seems to compile). what do you think? could you use this idea to avoid the repeated remove/insert and the clone in the error case?

```rust
        ...
        let mut res: HashMap<String, Vec<PartitionOffset>> = HashMap::with_capacity(n_topics);
        for (host, req) in reqs {
            let resp = try!(__send_receive::<_, protocol::OffsetResponse>(...));
            for tp in resp.topic_partitions {

                // re-write handler logic differently: first prepare the partition vector or fail if there's error thereby consuming the topic name
                let mut xs = Vec::with_capacity(tp.partitions.len());
                for p in tp.partitions {
                    match p.into_offset() {
                        Ok(x) => xs.push(x),
                        Err(e) => {
                            let code = match e { Error::Kafka(code) => code, _ => panic!("unknown error") };
                            return Err(Error::TopicPartitionError(tp.topic, p.partition, code));
                        }
                    }
                }
                // all fine, register the prepared offsets with the result structure
                res.insert(tp.topic, xs);
              }
            }
            Ok(res)
         }
```

**Comment by dead10ck** [comment]
Hmm, but isn't the whole reason the current code modifies the `res` object in place through the entry API because we could encounter more than one `TopicPartitionOffsetResponse` for the same topic? So if we just indiscriminately allocate a new `Vec` every iteration and `insert` it, then we could be overwriting previous data for that topic. We could allocate this `Vec` every iteration *and then* look up the current `Vec` stored inside the `res` object, but then we do an unnecessary allocation every time, which probably isn't much better than an unnecessary clone.

**Comment by dead10ck** [comment]
Perhaps another option is using `mem::replace` inside `into_offset` to conditionally take ownership of the topic `String` in the case of an error; we could pass in a `&mut String` and if there's an error, swap out the passed in value with the empty string (which is cheap because the empty `String` doesn't allocate anything), and put the newly-owned topic `String` into the returned error. It is kind of weird from an API perspective for the caller's `String` to get replaced, though. What do you think?

**Comment by xitep** [comment]
> [..] but isn't the whole reason the current code modifies the res object in place through the entry API that we could encounter more than one TopicPartitionOffsetResponse for the same topic [..]

oh yes, you are right! i overlooked the fact that we merge responses from different brokers and, indeed, my proposal would do the wrong thing just as you mentioned. and yes, the unnecessary extra allocation of a `Vec` for later merging is what we want to avoid in the first place. thanks for validating the idea.

i think i understand the `mem::replace` idea, and it sounds like it should work. but i would like if we could avoid it entirely. it would distribute the argument modifying logic between two remotely located methods which gets hard to maintain over time.

i just saw that the `HashMap::Entry` API has some methods which i was not aware of (in particular [remove_entry](https://doc.rust-lang.org/std/collections/hash_map/struct.OccupiedEntry.html#method.remove_entry) and [into_key](https://doc.rust-lang.org/std/collections/hash_map/struct.VacantEntry.html#method.into_key)) and it seems that these do allow now to gain ownership of the key within a hashmap. trying to fix up my previous idea, i came up with this verbose version (again not tested, but seems to compile):

```rust
        ...
        let mut res: HashMap<String, Vec<PartitionOffset>> = HashMap::with_capacity(n_topics);
        for (host, req) in reqs {
            let resp = try!(__send_receive::<_, protocol::OffsetResponse>(...));
            for tp in resp.topic_partitions {
                match res.entry(tp.topic) {
                    Entry::Occupied(mut e) => {
                        let partition_error = {
                            let mut partition_error = None;

                            // try pushing into xs which is hosted by
                            // the 'res' hashmap. if we encounter a
                            // partition error, we break out of the
                            // loop remembering we saw an error.
                            let xs = e.get_mut();
                            for p in tp.partitions {
                                match p.into_offset() {
                                    Ok(x) => xs.push(x),
                                    Err(e) => {
                                        let code = match e { Error::Kafka(code) => code, _ => panic!("unknown error") };
                                        partition_error = Some((p.partition, code));
                                        break;
                                    }
                                }                        
                            }
                            partition_error
                        };
                        // if we encountered a partition error, we
                        // take out the entry from the 'res' hashmap,
                        // thereby gaining ownership of the topic name
                        // and return with the appropriate error
                        if let Some((partition, error_code)) = partition_error {
                            let (topic, _) = e.remove_entry();
                            return Err(Error::TopicPartitionError(topic, partition, error_code));
                        }
                    }
                    Entry::Vacant(mut e) => {

                        // XXX basically the same as above but with a newly allocated 'xs'
                        
                    }
                }
         ...
        Ok(res)
}
```

here, we remove from the `res` hashmap only when we encounter an error (in which case we gain onwership of the original topic name without cloning), otherwise we keep pushing new entries into an existing `Vec`. this code is of course a very verbose, but i'd be happy if you could consider this approach (i hope there's a nice way to make this short) in favor of the magic with `mem::replace`.

do you want to have one more try? otherwise, i'd say your original version is good enough and we could go with that (the clone in the error case is actually ok, since it's only one per method call; regarding the repeated removal/insertion from/into the hashmap we'd need to measure of course, but it doesn't seem to me it's going to be a bottleneck; though it would be great if we could avoid this unnecessary work.)

**Comment by dead10ck** [comment]
@xitep ok, sorry it took so long, but I made some modifications to look similar to your example code above, which obviates the need for cloning the topic string, even in the error case :) let me know what you think!

File: src/client/mod.rs [hunk 11]
```diff
@@ -825,15 +850,66 @@ impl KafkaClient {
...
+                    // new collection to return.
+                    let resp_offsets = match entry {
+                        hash_map::Entry::Occupied(ref mut e) => e.get_mut(),
+                        hash_map::Entry::Vacant(_) => {
+                            new_resp_offsets = Some(Vec::new());
+
+                            match new_resp_offsets {
+                                Some(ref mut v) => v,
+                                None => panic!("Option value was None after Some assignment"),
+                            }
```
Comment (xitep) [inline src/client/mod.rs]: i think it's fine to shorten this to a mere:

```rust
... {
  new_resp_offsets = Some(Vec::new());
  new_resp_offsets.as_mut().unwrap()
}
```

File: src/client/mod.rs [hunk 12]
```diff
@@ -825,15 +850,66 @@ impl KafkaClient {
...
+                    let topic = KafkaClient::get_key_from_entry(entry);
+                    return Err(Error::TopicPartitionError(topic, partition, code));
+                }
+
+                if let hash_map::Entry::Vacant(e) = entry {
+                    // unwrap is ok because if it is Vacant, it would have
+                    // been made into a Some above
+                    e.insert(new_resp_offsets.unwrap());
                 }
             }
```
Comment (xitep) [inline src/client/mod.rs]: may i ask you to remove the empty lines within this `for tp in resp.topic_partitions {` loop?

File: src/client/mod.rs [hunk 13]
```diff
@@ -825,15 +850,66 @@ impl KafkaClient {
...
             for tp in resp.topic_partitions {
-                let e = res.entry(tp.topic).or_insert(vec![]);
-                for p in tp.partitions {
-                    e.push(p.into_offset());
+                let mut entry = res.entry(tp.topic);
+                let mut new_resp_offsets = None;
+                let mut err = None;
+
+                // Use an explicit scope here to allow insertion into a vacant entry
+                // below
```
Comment (xitep) [inline src/client/mod.rs]: no need for this comment. the borrow checker will hit us immediately when we try not to use the extra scope :)

**Comment by xitep** [comment]
@dead10ck this is wonderful! i can imagine you fought the borrowck a lot but in the end you made it :) that's quite cool :sunglasses:

i have had just some minor cosmetics remarks; but it seems to me we're ready to merge.

File: src/client/mod.rs (hunk 11 above)
Comment (dead10ck) [inline src/client/mod.rs]: Ah, nice; I just learned `as_mut` existed :) will fix

File: src/client/mod.rs (hunk 12 above)
Comment (dead10ck) [inline src/client/mod.rs]: Sure, I can do that. I personally find whitespace between blocks improves readability, but I agree it should be consistent with the rest of the crate's style.

**Comment by dead10ck** [comment]
@xitep ok, I've removed the whitespace and did a rebase to only a few commits. If it looks good to you, I think it's ready to merge.

**Comment by xitep** [comment]
thank you for all the hard work!
